runtime/src/main/java/com/datastax/oss/quarkus/runtime/internal/health/CassandraAsyncHealthCheck.java

```diff
@@ -0,0 +1,104 @@
+
+  private static final Type COMPLETION_STAGE_OF_QUARKUS_CQL_SESSION_TYPE =
+      new TypeLiteral<CompletionStage<QuarkusCqlSession>>() {}.getType();
+  static final String HEALTH_CHECK_QUERY =
+      "SELECT data_center, release_version, cluster_name, cql_version FROM system.local";
+
+  private CompletionStage<QuarkusCqlSession> cqlSessionCompletionStage;
+
+  @SuppressWarnings("unchecked")
+  public CompletionStage<QuarkusCqlSession> beanProvider() {
+    return (CompletionStage<QuarkusCqlSession>)
+        Arc.container().instance(COMPLETION_STAGE_OF_QUARKUS_CQL_SESSION_TYPE).get();
+  }
+
+  @PostConstruct
+  protected void init() {
+    this.cqlSessionCompletionStage = beanProvider();
+  }
+
+  @Override
+  public Uni<HealthCheckResponse> call() {
+    return Uni.createFrom()
+        .completionStage(
+            cqlSessionCompletionStage.thenCompose(
+                cqlSession ->
+                    cqlSession
+                        .executeAsync(HEALTH_CHECK_QUERY)
+                        .thenApply(AsyncPagingIterable::one)
+                        .thenApply(
+                            result -> {
+                              HealthCheckResponseBuilder builder =
+                                  HealthCheckResponse.named(HEALTH_CHECK_NAME).up();
+                              if (result == null) {
+                                return builder
+                                    .down()
+                                    .withData("reason", "system.local returned null")
+                                    .build();
+                              }
+                              for (Map.Entry<String, String> entry :
+                                  extractInfoFromResult(result).entrySet()) {
+                                builder.withData(entry.getKey(), entry.getValue());
+                              }
+                              return builder
+                                  .withData(
+                                      "numberOfNodes", cqlSession.getMetadata().getNodes().size())
+                                  .up()
+                                  .build();
+                            })
+                        .exceptionally(
+                            ex -> {
+                              HealthCheckResponseBuilder builder =
+                                  HealthCheckResponse.named(HEALTH_CHECK_NAME).up();
+                              return builder.down().withData("reason", ex.getMessage()).build();
+                            })));
+  }
+
+  private Map<String, String> extractInfoFromResult(Row result) {
+    HashMap<String, String> details = new HashMap<>();
+    details.put("datacenter", result.getString("data_center"));
+    details.put("releaseVersion", result.getString("release_version"));
+    details.put("clusterName", result.getString("cluster_name"));
+    details.put("cqlVersion", result.getString("cql_version"));
+    return details;
+  }
```

Comment:
Nit: it's probably slightly better to specify the list of columns we wish to query and extract in a collection somewhere.  We can then use that collection to build HEALTH_CHECK_QUERY as well as iterate over the same collection here when we wish to extract values.  Advantage is that we specify those columns in a single spot and re-use them in both operations.

Reply:
same as here:
https://github.com/datastax/cassandra-quarkus/pull/127#discussion_r482764654